### PR TITLE
techtask/Separation-of-credentials-from-endpoints 

### PR DESCRIPTION
**Pull request by yahorm**

### CODE REVIEW

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt [hunk 1]
```diff
@@ -62,18 +65,38 @@ class ConfigManager : AbstractVerticle() {
...
+        return array.stream()
+                .map { it as JsonObject }
+                .filter {
+                    id == it.getString(ID)
+                }.findFirst()
+    }
+
     companion object {
         val LOGGER: Logger = LoggerFactory.getLogger(ConfigManager::class.java)
+        const val ID = "id";
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt]: Nice i like that

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt [hunk 2]
```diff
@@ -62,18 +65,38 @@ class ConfigManager : AbstractVerticle() {
...
+        endpoint.remove(CREDENTIALS).toString().let {
+            val credentials = findJsonObjectById(it, credentials)
+                    .orElse(JsonObject())
+            endpoint.put(USER, credentials.getString(USER))
+            endpoint.put(PASSWORD, credentials.getString(PASSWORD))
+        }
+        return endpoint
+    }
+
+    private fun findJsonObjectById(id: String?, array: JsonArray): Optional<JsonObject> {
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt]: You can use **Extension Functions** this.
https://kotlinlang.org/docs/reference/extensions.html#extension-functions

**Comment by michalskrzypek** [comment]
Refactored code and added unit tests

**Comment by michalskrzypek** [comment]
> Let's introduce JUnit tests with this improvement.

Done. Code refactored and unit tests added.

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt [hunk 3]
```diff
@@ -10,15 +10,20 @@ import io.vertx.core.json.JsonArray
...
 
     private val widgets = mutableMapOf<String, Widget>()
     private lateinit var storage: Storage
     private lateinit var endpoints: JsonArray
+    private lateinit var credentials: JsonArray
+    private lateinit var endpoint: Endpoint
 
     override fun start() {
-        endpoints = config().getJsonArray("endpoints")
+        endpoints = config().getJsonArray(ENDPOINTS)
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt]: this is know confusing - in my opinion there is no need to store `endpoints`, `credentials` and most importantly `endpoint` here. Look below to see how I see it:

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt [hunk 4]
```diff
@@ -78,18 +83,12 @@ class ConfigManager : AbstractVerticle() {
...
         val endpointId = config.getString(CogboardConstants.PROP_ENDPOINT)
         endpointId?.let {
-            config.put(CogboardConstants.PROP_ENDPOINT, endpoints
-                    .stream()
-                    .map { it as JsonObject }
-                    .filter {
-                        endpointId == it.getString("id")
-                    }.findFirst()
-                    .orElse(JsonObject()))
+            config.put(CogboardConstants.PROP_ENDPOINT, endpoint.readById(endpointId))}
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt]: I would like to see it like this:
`config.put(CogboardConstants.PROP_ENDPOINT, Endpoint.from(config(), endpointId))`

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt [hunk 5]
```diff
@@ -78,18 +83,12 @@ class ConfigManager : AbstractVerticle() {
...
         }
-    }
 
     companion object {
         val LOGGER: Logger = LoggerFactory.getLogger(ConfigManager::class.java)
+        const val CREDENTIALS = "credentials"
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigManager.kt]: This should be moved to `Endpoint` class

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/handler/GetEndpoints.kt [hunk 6]
```diff
@@ -24,8 +24,8 @@ class GetEndpoints : RoutingHandlerFactory {
         copy.stream().forEach {
             if (it is JsonObject) {
                 it.remove("url")
-                it.remove("user")
-                it.remove("password")
+                it.remove("publicUrl")
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/handler/GetEndpoints.kt]: Nice !

File: cogboard-app/src/test/kotlin/com/cognifide/cogboard/config/EndpointTest.kt [hunk 7]
```diff
@@ -0,0 +1,87 @@
+package com.cognifide.cogboard.config
+
+import io.vertx.core.json.JsonArray
+import io.vertx.core.json.JsonObject
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertFalse
+import org.junit.jupiter.api.Test
+
+internal class EndpointTest {
```
Comment (szymon-owczarzak) [inline cogboard-app/src/test/kotlin/com/cognifide/cogboard/config/EndpointTest.kt]: I really like that You introduced JUnit to the project !

File: cogboard-app/src/test/resources/com/cognifide/cogboard/config/credentials-test.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
+[
```
Comment (szymon-owczarzak) [inline cogboard-app/src/test/resources/com/cognifide/cogboard/config/credentials-test.json]: merge `credentials-test.json` and `endpoints-test.json` so it is presented as in actual code.

File: cogboard-app/src/test/kotlin/com/cognifide/cogboard/config/EndpointTest.kt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    private val endpoints: JsonArray = readJsonArrayFromResource("/com/cognifide/cogboard/config/endpoints-test.json", ENDPOINTS)
+    private val credentials: JsonArray = readJsonArrayFromResource("/com/cognifide/cogboard/config/endpoints-test.json", CREDENTIALS)
+
+    @Test
+    fun shouldRemoveCredentialsProp() {
+        val validEndpoint = Endpoint(endpoints, credentials,"validEndpoint").asJson()
```
Comment (yahorm) [inline cogboard-app/src/test/kotlin/com/cognifide/cogboard/config/EndpointTest.kt]: next time please use @Before for initialize e.g. endpoint once.

```   
private lateinit var validEndpoint: JsonObject
private lateinit var invalidEndpoint: JsonObject

@BeforeEach
fun init() {
    validEndpoint = Endpoint(endpoints, credentials, "validEndpoint").asJson()
    invalidEndpoint = Endpoint(endpoints, credentials, "invalidEndpoint").asJson()
}
```

**Pull request merged**